src/vmm/src/vstate/system.rs

```diff
@@ -50,7 +50,12 @@ pub struct KvmContext {
 impl KvmContext {
     pub fn new() -> Result<Self> {
         use kvm_ioctls::Cap::*;
-        let kvm = Kvm::new().expect("Error creating the Kvm object");
+        let kvm = Kvm::new().expect(
+            "Critical error creating the KVM object. \
```

Comment:
Not sure why we expect here and not gracefully handle the error (not having the right capabilities isn t less bad than this). So I suggest doing a `map_err()?`instead of `expect`. 
If we do that, we can also pretty print the `Error (13)` cryptical message.

Reply:
Thanks for the suggestion. I will make the change but clarify this would permit firecracker to be used with limited features with the user configuration as is.

For addressing the immediate issue raised this change seems like a reasonable improvement. But it isn't clear to me yet how KVM permits memory and vCPU usage but it fails with network configuration, even without the ACL configured for `/dev/kvm`. Are there other spots where this missing ACL could potentially cause a "panic"?

Reply:
I am not sure I understand what you mean by KVM failing with network configuration? I agree that this is a reasonable improvement and we can go ahead with current version

Reply:
> I am not sure I understand what you mean by KVM failing with network configuration?

To reproduce the [issue](https://github.com/firecracker-microvm/firecracker/issues/2908), I had to (via the API):
1) Configure the kernel
1) Configure the file system
1) *Configure the network (tap) device for the guest*
1) Launch the guest instance

If the step to configure the network device is omitted, the instance will launch successfully. (at least in my initial attempt to duplicate the issue).

Unfortunately using a `map_err()` will leave the kvm reference that is supposed to be created will be null. I'm inclined to use `express()` for now to force the exit of the process at this stage. At the very least the user will get a clearer message as to what is wrong.

Reply:
> If the step to configure the network device is omitted, the instance will launch successfully. (at least in my initial attempt to duplicate the issue).

Can you retry this? It doesn't have something to do with the network device.
I remember I bumped into this error even without a tap device configured

Reply:
I suggest you add a new variant for the `Error` enum in `src/vmm/vstate/system.rs`, like `KvmCreateVm(errno::Error)` and implement the `Display` trait for it such that you output the desired error message, together with the wrapped error.
Also do a check to see whether the `Debug` or `Display` implementations are used higher up in the calling chain

Reply:
The issue is not related to network in any way. You can replicate it by running Firecracker with a user that does not have access to `/dev/kvm`. The `Error(13)` is debug code for "Permission Denied" error.

Reply:
> I suggest you add a new variant for the `Error` enum in `src/vmm/vstate/system.rs`, like `KvmCreateVm(errno::Error)` and implement the `Display` trait for it such that you output the desired error message, together with the wrapped error. Also do a check to see whether the `Debug` or `Display` implementations are used higher up in the calling chain

Since we also want to force process exit and also pretty print the kvm_ioctls error, we can use:
`option.unwrap_or_else(|e| panic!("Cannot create KVM object: {}", e))`

Reply:
> If the step to configure the network device is omitted, the instance will launch successfully. (at least in my initial attempt to duplicate the issue).
>> Can you retry this? It doesn't have something to do with the network device.
You are correct. While I have got two results on two machines, I have confirmed at least once to be a problem that was not net config specific.

While the forced process exit works, I think defining a new error enum entry with a tidy formatted display is reasonable in the context of this PR.

Reply:
> Since we also want to force process exit and also pretty print the kvm_ioctls error, we can use:
option.unwrap_or_else(|e| panic!("Cannot create KVM object: {}", e))

Do we neccessarily want to force process exits?
If we were to use regular error propagation using `?`, it would either result in an API error for the `InstanceStart` call or a process exit if Firecracker was started without an API thread.

I think this is perfectly ok (and preferred actually), since a user may then add the right permission to `/dev/kvm` and retry the API call.

@dianpopa @mattschlebusch WDYT?

Reply:
My current thinking here is that it is possible that permissions to KVM are organized by a user group instead of the ACL in which case the FC process would need to get restarted anyway when the user starts a new session for the group's membership to take effect. An argument could be made either way though.

Reply:
If I can interpret the principal that you state about not closing the process and the customer is in control of the process makes sense. I'm happy to use that as the metric for the decision.